scanner/src/main/kotlin/experimental/PackageProvenanceResolver.kt

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (C) 2021 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.experimental
+
+import java.io.IOException
+import java.net.HttpURLConnection
+
+import okhttp3.Request
+
+import org.ossreviewtoolkit.downloader.VersionControlSystem
+import org.ossreviewtoolkit.model.ArtifactProvenance
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.Provenance
+import org.ossreviewtoolkit.model.RemoteArtifact
+import org.ossreviewtoolkit.model.RepositoryProvenance
+import org.ossreviewtoolkit.model.SourceCodeOrigin
+import org.ossreviewtoolkit.model.UnknownProvenance
+import org.ossreviewtoolkit.model.VcsInfo
+import org.ossreviewtoolkit.utils.OkHttpClientHelper
+import org.ossreviewtoolkit.utils.collectMessagesAsString
+import org.ossreviewtoolkit.utils.createOrtTempDir
+import org.ossreviewtoolkit.utils.log
+import org.ossreviewtoolkit.utils.showStackTrace
+
+/**
+ * The [PackageProvenanceResolver] provides a function to resolve the [Provenance] of a [Package].
+ */
+interface PackageProvenanceResolver {
```

Comment:
I'm having a bit of a problem with the term "resolve" here. To me, "resolve" implies working on something that *is* already there, but gets "transformed" in some way (like resolving a link to the file pointed to). However, what I believe this does, is to *determine* the provenance. Or even "create" / "calculate" the provenance.

Reply:
IMO the term "resolve" is okay, but others would work as well. Based on your interpretation, it does take something which is already there, namely the package metadata, and transforms it into a provenance. For example the VCS info gets amended with a resolved revision.

Reply:
What I mean is: In "resolve" there is a 1:1 relationship between input and output, you always resolve *to* the same thing. But here, you determine the provenance from metadata, and what the provenance *is* changes depending on the `sourceCodeOriginPriority`. I'd strongly prefer another term, also to avoid the confusion with "resolving revisions" similar to like we had before in the `VcsInfo` context; otherwise one could think this somehow resolves the (symbolic) revision of a `RepositoryProvenance`.

Reply:
Ok, I'd suggest one of those solutions:
* Please suggest a specific name and I'll change it.
* We'll leave the naming discussion for later.
Just because I don't want to get stuck on a naming discussion right now. My current plan is to try to get a working first version of the new scanner ready before my holiday and then maybe after the holiday we could make a extensive review session for the concept and implementation.

Reply:
I was giving the naming some thought again, but couldn't come up with a better name. And as the existence *is* actually checked (e.g. via a `HEAD` request in case of a source artifact), the artifact *is* actually implicitly resolved. So let's keep the naming as-is then for now.